hanish/bot.py

```diff
@@ -0,0 +1,409 @@
+# hanish.bot
+# Primary interface to the hanish Slack bot.
+#
+# Author:   Benjamin Bengfort <benjamin@bengfort.com>
+# Created:  Mon May 08 14:44:06 2017 -0400
+#
+# Copyright (C) 2016 Bengfort.com
+# For license information, see LICENSE.txt
+#
+# ID: bot.py [] benjamin@bengfort.com $
+
+"""
+Primary interface to the hanish Slack bot. The bot listens for chat messages
+and responds appropriately, also looking up weather by zip code.
+
+See also: https://www.fullstackpython.com/blog/build-first-slack-bot-python.html
+"""
+
+##########################################################################
+## Imports
+##########################################################################
+
+import os
+import re
+import json
+import time
+import signal
+import schedule
+
+from .chat import *
+from .exceptions import *
+from .utils import memoized
+from .darksky import DarkSky
+from .zipcode import ZipCodeDB
+from slackclient import SlackClient
+
+# time to wait between websocket reads
+POLLING_INTERVAL = 1
+
+
+##########################################################################
+## Helper Functions
+##########################################################################
+
+def environ_default(kwargs, name, envvar, default=None, required=False):
+    """
+    This helper function fetches a default value from the specified envvar if
+    the name doesn't exist in the kwargs dictionary. If the name or envvar are
+    not defined, then this function returns the default value unles required
+    is True, at which point an ImproperlyConfigured exception is raised.
+    """
+
+    # If the name is in the keyword arguments, return its value
+    if name in kwargs: return kwargs[name]
+
+    # If the envvar is in the environment, return its value.
+    if envvar in os.environ: return os.environ[envvar]
+
+    # If required is True and we're here, raise an exception.
+    if required:
+        raise ImproperlyConfigured(
+            "Required config {} not found in ${}".format(name, envvar)
+        )
+
+    # Otherwise return the default
+    return default
+
+
+##########################################################################
+## Hanish Bot
+##########################################################################
+
+class Bot(object):
```

Comment:
I like the idea of decoupling the logic from the side effect. Let's do it in Haskell!